### PR TITLE
Emit key events from Scene3D

### PR DESCRIPTION
**Pull request by ahcorde**
Signed-off-by: ahcorde <ahcorde@gmail.com>

# 🎉 New feature

## Summary

This PR builds on top of this other PR https://github.com/ignitionrobotics/ign-gui/pull/213. This PR makes the 3D scene emit key events (press and release), that can be used by other plugins. 

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

File: include/ignition/gui/GuiEvents.hh
Comment (mjcarroll) [inline include/ignition/gui/GuiEvents.hh:296]: Where does this `-9` come from?  Seems kind of magic.

File: src/plugins/scene3d/Scene3D.cc
Comment (mjcarroll) [inline src/plugins/scene3d/Scene3D.cc:1680]: Curious why this is needed?  Aren't you static casting to/from the same type here?

File: include/ignition/gui/GuiEvents.hh
Comment (ahcorde) [reply include/ignition/gui/GuiEvents.hh:296]: The first event type starts with `0`, then we are subtracting one to each new event

Render:
https://github.com/ignitionrobotics/ign-gui/blob/a983a6d7e39473ee69ebf505b7b82f04b8d7c605/include/ignition/gui/GuiEvents.hh#L45

SnapIntervals:
https://github.com/ignitionrobotics/ign-gui/blob/a983a6d7e39473ee69ebf505b7b82f04b8d7c605/include/ignition/gui/GuiEvents.hh#L87

File: src/plugins/scene3d/Scene3D.cc
Comment (ahcorde) [reply src/plugins/scene3d/Scene3D.cc]: https://github.com/ignitionrobotics/ign-gui/pull/224/commits/54ec72537a6a7fc75620c22f4019efa6eb39dd17

File: include/ignition/gui/GuiEvents.hh
Comment (scpeters) [inline include/ignition/gui/GuiEvents.hh:298]: "Get the key within the scene that the user pressed"

File: src/plugins/scene3d/Scene3D.hh
Comment (scpeters) [inline src/plugins/scene3d/Scene3D.hh:151]: `release` -> `press`

**Comment by ahcorde** [comment]
Do not merge this PR yet, i will add key modifiers such us control, shift or alt.

**Comment by ahcorde** [comment]
I think we should add the same modifiers to the Mouse clicks, what do you think @scpeters ?

**Comment by ahcorde** [comment]
@scpeters and @mjcarroll this is ready for another pass

**Comment by mjcarroll** [comment]
Looks like something has changed on `focal`, presumably because of Qt5?

```
  /github/workspace/test/integration/scene3d.cc:271: Failure
  Expected equality of these values:
    Qt::Key_A
      Which is: 65
    keyPressedValue
      Which is: 16777251
 ```

**Comment by ahcorde** [comment]
@mjcarroll  I think it's flaky...

**Comment by mjcarroll** [comment]
I think adding the copy constructor messed with all of the derived classes here as well.

File: src/plugins/scene3d/Scene3D.cc
Comment (chapulina) [inline src/plugins/scene3d/Scene3D.cc:1066]: Nit: this can be moved into the `if`, right? No need to create the object if we aren't going to use it. The same below

File: src/plugins/scene3d/Scene3D.hh
Comment (chapulina) [inline src/plugins/scene3d/Scene3D.hh:148]:
```suggestion
    /// \brief Broadcasts the current key release
```

File: src/plugins/scene3d/Scene3D.hh
Comment (chapulina) [inline src/plugins/scene3d/Scene3D.hh:151]:
```suggestion
    /// \brief Broadcasts the current key press
```

**Review by chapulina — CHANGES_REQUESTED** [review]
The idea of the PR is good, we just need to fix the build:

```
   In file included from /github/workspace/src/GuiEvents_TEST.cc:21:
  /github/workspace/include/ignition/gui/GuiEvents.hh: In constructor 'ignition::gui::events::KeyRelease::KeyRelease(const ignition::common::KeyEvent&)':
  /github/workspace/include/ignition/gui/GuiEvents.hh:269:38: error: use of deleted function 'ignition::common::KeyEvent::KeyEvent(const ignition::common::KeyEvent&)'
               : QEvent(kType), key(_key)
```

**Comment by ahcorde** [comment]
@chapulina PR is in https://github.com/ignitionrobotics/ign-common/pull/224

We need a `ign-common3` release.

File: include/ignition/gui/GuiEvents.hh
Comment (chapulina) [inline include/ignition/gui/GuiEvents.hh:264]: Can these have the `OnScene` suffix like on #228 for clarity?

Comment (ahcorde) [reply include/ignition/gui/GuiEvents.hh]: Done https://github.com/ignitionrobotics/ign-gui/pull/224/commits/906b0dd84176f35c9627b5140a426c86b2e86573

**Review by mjcarroll — APPROVED** [review]

**Comment by chapulina** [comment]
Same as #228, since this is not used by any Citadel plugins yet, let's target it at `main`.

My concern with these events is that they're not PIMPL'ed and once merged into a stable version we can't iterate on them (see #239 ).

**Comment by ahcorde** [comment]
@chapulina retargeted to `main` and add PIMPL pattern

**Review by chapulina — APPROVED** [review]
Pushed tweaks on cdcf459ffa932386d78b1f61ad98aa391f262b06.

LGTM with happy CI.

**Comment by ahcorde** [comment]
Merging, some know tests are failing: `UNIT_WorldStats_TEST` and `UNIT_Application_TEST`

**Comment by ahcorde** [comment]
@osrf-jenkins retest this please

**Comment by ahcorde** [comment]
@osrf-jenkins retest this please

**Pull request merged**
